{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "%matplotlib notebook\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import networkx as nx\n",
    "import os\n",
    "from IPython.display import Image\n",
    "import seqalign\n",
    "import reveal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Experiment 3 - Creating and aligning diploid contig representations "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We used falcon to construct an assembly of a diploid model organism, C. Elegans. This resulted in three files:\n",
    "- p_ctg.fa, containing the primary contigs\n",
    "- a_ctg_base.fa, containing the base contigs\n",
    "- a_ctg_all.fa, containing the alternative/associated contigs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We wrote a small script that takes these three files and uses REVEAL to align the associated contigs on top of the primary/base contigs in order to construct graph structure for each contig, in which the two homologous chromosomes of the sequenced organism are described by different paths through the graph. We don't know which path corresponds to which chromosome, but, this way, at least we can account for some of the variation observed between the two chromosomes in further downstream analysis. Let's obtain these graph structures."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "!falcon2gfa.py p_ctg.fa a_ctg_base.fa a_ctg_all.fa --align -m5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From these graphs directly we can already detect all variations between the two chromosomes. Lets do that. For now, only consider the graphs larger than 1MB, to save some time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 139,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "contigs=!find . -type f -size +1M -name \"0000*.gfa\" -exec basename \\{\\} \\;"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for gfa in contigs:\n",
    "    !reveal bubbles {gfa} > {gfa.replace(\".gfa\",\"\")}.bubbles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So now, we created a single file for eacht contig describing the heterozygous sequence variations that follow from aligning the associated contigs to the primary contig. Let's calculate some statistics on these variations. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected 150 heterozygous SNPs.\n",
      "Detected 6200 heterozygous indels.\n",
      "Detected 77 heterozygous structural variants\n"
     ]
    }
   ],
   "source": [
    "_hetsnps=[]\n",
    "_hetindels=[]\n",
    "_hetstruct=[]\n",
    "\n",
    "multiallele=0\n",
    "complexb=0\n",
    "for gfa in contigs:\n",
    "    bubblefile=gfa.replace(\".gfa\",\".bubbles\")\n",
    "    with open(bubblefile) as bubbles:\n",
    "        header=bubbles.readline()\n",
    "        headers=header.split()\n",
    "        altidx1=6\n",
    "        altidx2=7\n",
    "        \n",
    "        for line in bubbles:\n",
    "            bubble=line.split()\n",
    "            alleles=bubble[5].split(',')\n",
    "            \n",
    "            if len(alleles)>2:\n",
    "                multiallele+=1\n",
    "                continue #more than 2 alleles, skip\n",
    "            if 'N' in alleles:\n",
    "                complexb+=1\n",
    "                continue #complex bubble\n",
    "            \n",
    "            varlengths=[len(a) for a in alleles]\n",
    "            size=max(varlengths)\n",
    "            snp=set([1])==set(varlengths) and '-' not in alleles\n",
    "            indel='-' in alleles\n",
    "            intersectsall=True#bubble[refidx]!='-' and bubble[altidx1]!='-' and bubble[altidx2]!='-'\n",
    "            \n",
    "            if snp and intersectsall:\n",
    "                if bubble[altidx1]!=bubble[altidx2]:\n",
    "                    _hetsnps.append((bubblefile,line))\n",
    "            if indel and intersectsall:\n",
    "                if bubble[altidx1]!=bubble[altidx2]:\n",
    "                    _hetindels.append((bubblefile,line))\n",
    "                    \n",
    "            if size>50 and intersectsall:\n",
    "                if bubble[altidx1]!=bubble[altidx2]:\n",
    "                    _hetstruct.append((bubblefile,line))\n",
    "\n",
    "print \"Detected %d heterozygous SNPs.\"%len(_hetsnps)\n",
    "print \"Detected %d heterozygous indels.\"%len(_hetindels)\n",
    "print \"Detected %d heterozygous structural variants\"%len(_hetstruct)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "So, we mainly see indel variations, shorter than 50bp. Which is to be expected, considering that we did not polish the assembly and the main sequencing error in SMRT sequencing are short indels."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, that we have our graphs we can use REVEAL to align them to the WBcel235 haploid assembly."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Lets download the reference genome and orient all contigs/graphs larger than 1MB before aligning them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2016-07-27 16:46:59--  ftp://ftp.ensemblgenomes.org/pub/metazoa/release-31/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.31.dna.chromosome.III.fa.gz\n",
      "           => 'Caenorhabditis_elegans.WBcel235.31.dna.chromosome.III.fa.gz'\n",
      "Resolving ftp.ensemblgenomes.org... 193.62.197.94\n",
      "Connecting to ftp.ensemblgenomes.org|193.62.197.94|:21... connected.\n",
      "Logging in as anonymous ... Logged in!\n",
      "==> SYST ... done.    ==> PWD ... done.\n",
      "==> TYPE I ... done.  ==> CWD (1) /pub/metazoa/release-31/fasta/caenorhabditis_elegans/dna ... done.\n",
      "==> SIZE Caenorhabditis_elegans.WBcel235.31.dna.chromosome.III.fa.gz ... 4125615\n",
      "==> PASV ... done.    ==> RETR Caenorhabditis_elegans.WBcel235.31.dna.chromosome.III.fa.gz ... done.\n",
      "Length: 4125615 (3.9M) (unauthoritative)\n",
      "\n",
      "Caenorhabditis_eleg 100%[=====================>]   3.93M  10.1MB/s   in 0.4s   \n",
      "\n",
      "2016-07-27 16:46:59 (10.1 MB/s) - 'Caenorhabditis_elegans.WBcel235.31.dna.chromosome.III.fa.gz' saved [4125615]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!wget ftp://ftp.ensemblgenomes.org/pub/metazoa/release-31/fasta/caenorhabditis_elegans/dna/Caenorhabditis_elegans.WBcel235.31.dna.genome.fa.gz\n",
    "!gzip -d Caenorhabditis_elegans.WBcel235.31.dna.genome.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "07/31/2016 12:02:47 AM Processing 000000F.gfa\n",
      "07/31/2016 12:03:55 AM Not reverse complementing 000000F.gfa\n",
      "07/31/2016 12:03:55 AM Processing 000001F.gfa\n",
      "07/31/2016 12:05:04 AM Reverse complementing 000001F.gfa\n",
      "07/31/2016 12:05:04 AM Processing 000002F.gfa\n",
      "07/31/2016 12:06:06 AM Reverse complementing 000002F.gfa\n",
      "07/31/2016 12:06:06 AM Processing 000003F.gfa\n",
      "07/31/2016 12:07:10 AM Not reverse complementing 000003F.gfa\n",
      "07/31/2016 12:07:10 AM Processing 000004F.gfa\n",
      "07/31/2016 12:08:21 AM Reverse complementing 000004F.gfa\n",
      "07/31/2016 12:08:21 AM Processing 000005F.gfa\n",
      "07/31/2016 12:09:26 AM Not reverse complementing 000005F.gfa\n",
      "07/31/2016 12:09:26 AM Processing 000006F.gfa\n",
      "07/31/2016 12:10:26 AM Reverse complementing 000006F.gfa\n",
      "07/31/2016 12:10:26 AM Processing 000007F.gfa\n",
      "07/31/2016 12:11:25 AM Not reverse complementing 000007F.gfa\n",
      "07/31/2016 12:11:25 AM Processing 000008F.gfa\n",
      "07/31/2016 12:12:25 AM Not reverse complementing 000008F.gfa\n",
      "07/31/2016 12:12:25 AM Processing 000009F.gfa\n",
      "07/31/2016 12:13:30 AM Not reverse complementing 000009F.gfa\n",
      "07/31/2016 12:13:30 AM Processing 000010F.gfa\n",
      "07/31/2016 12:14:31 AM Reverse complementing 000010F.gfa\n",
      "07/31/2016 12:14:31 AM Processing 000011F.gfa\n",
      "07/31/2016 12:15:28 AM Reverse complementing 000011F.gfa\n",
      "07/31/2016 12:15:28 AM Processing 000012F.gfa\n",
      "07/31/2016 12:16:26 AM Not reverse complementing 000012F.gfa\n",
      "07/31/2016 12:16:26 AM Processing 000013F.gfa\n",
      "07/31/2016 12:17:21 AM Reverse complementing 000013F.gfa\n",
      "07/31/2016 12:17:21 AM Processing 000014F.gfa\n",
      "07/31/2016 12:18:17 AM Reverse complementing 000014F.gfa\n",
      "07/31/2016 12:18:17 AM Processing 000015F.gfa\n",
      "07/31/2016 12:19:12 AM Reverse complementing 000015F.gfa\n",
      "07/31/2016 12:19:12 AM Processing 000016F.gfa\n",
      "07/31/2016 12:20:07 AM Not reverse complementing 000016F.gfa\n",
      "07/31/2016 12:20:07 AM Processing 000017F.gfa\n",
      "07/31/2016 12:21:02 AM Reverse complementing 000017F.gfa\n",
      "07/31/2016 12:21:02 AM Processing 000018F.gfa\n",
      "07/31/2016 12:21:58 AM Reverse complementing 000018F.gfa\n",
      "07/31/2016 12:21:58 AM Processing 000019F.gfa\n",
      "07/31/2016 12:22:52 AM Not reverse complementing 000019F.gfa\n",
      "07/31/2016 12:22:52 AM Processing 000020F.gfa\n",
      "07/31/2016 12:23:47 AM Not reverse complementing 000020F.gfa\n",
      "07/31/2016 12:23:47 AM Processing 000021F.gfa\n",
      "07/31/2016 12:24:42 AM Reverse complementing 000021F.gfa\n",
      "07/31/2016 12:24:42 AM Processing 000022F.gfa\n",
      "07/31/2016 12:25:36 AM Reverse complementing 000022F.gfa\n",
      "07/31/2016 12:25:36 AM Processing 000023F.gfa\n",
      "07/31/2016 12:26:32 AM Not reverse complementing 000023F.gfa\n",
      "07/31/2016 12:26:32 AM Processing 000024F.gfa\n",
      "07/31/2016 12:27:26 AM Not reverse complementing 000024F.gfa\n",
      "07/31/2016 12:27:26 AM Processing 000025F.gfa\n",
      "07/31/2016 12:28:20 AM Not reverse complementing 000025F.gfa\n",
      "07/31/2016 12:28:20 AM Processing 000026F.gfa\n",
      "07/31/2016 12:29:14 AM Reverse complementing 000026F.gfa\n",
      "07/31/2016 12:29:14 AM Processing 000028F.gfa\n",
      "07/31/2016 12:30:09 AM Not reverse complementing 000028F.gfa\n",
      "07/31/2016 12:30:09 AM Processing 000029F.gfa\n",
      "07/31/2016 12:31:03 AM Reverse complementing 000029F.gfa\n",
      "07/31/2016 12:31:03 AM Processing 000031F.gfa\n",
      "07/31/2016 12:31:57 AM Not reverse complementing 000031F.gfa\n",
      "07/31/2016 12:31:57 AM Processing 000032F.gfa\n",
      "07/31/2016 12:32:50 AM Not reverse complementing 000032F.gfa\n",
      "07/31/2016 12:32:50 AM Processing 000034F.gfa\n",
      "07/31/2016 12:33:45 AM Reverse complementing 000034F.gfa\n",
      "07/31/2016 12:33:45 AM Processing 000035F.gfa\n",
      "07/31/2016 12:34:38 AM Reverse complementing 000035F.gfa\n",
      "07/31/2016 12:34:38 AM Processing 000036F.gfa\n",
      "07/31/2016 12:35:32 AM Reverse complementing 000036F.gfa\n",
      "07/31/2016 12:35:32 AM Processing 000037F.gfa\n",
      "07/31/2016 12:36:25 AM Not reverse complementing 000037F.gfa\n"
     ]
    }
   ],
   "source": [
    "!reveal orient Caenorhabditis_elegans.WBcel235.31.dna.genome.fa {\" \".join(contigs)}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that all graphs are oriented with respect to the reference genome, we can align them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "07/31/2016 12:36:28 AM Reading graph: 000000F.gfa ...\n",
      "07/31/2016 12:36:29 AM Done.\n",
      "07/31/2016 12:36:29 AM Constructing index...\n",
      "07/31/2016 12:36:55 AM Done.\n",
      "07/31/2016 12:36:55 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:37:23 AM Merging nodes...\n",
      "07/31/2016 12:37:23 AM Done.\n",
      "07/31/2016 12:37:23 AM Done (99.41% identity, 4645234 bases out of 4672659 aligned, 6218 nodes out of 12579 aligned).\n",
      "07/31/2016 12:37:23 AM Writing graph...\n",
      "07/31/2016 12:37:25 AM Done.\n",
      "07/31/2016 12:37:25 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000000F.gfa\n",
      "07/31/2016 12:37:28 AM Reading graph: 000001F.gfa ...\n",
      "07/31/2016 12:37:28 AM Done.\n",
      "07/31/2016 12:37:28 AM Constructing index...\n",
      "07/31/2016 12:37:55 AM Done.\n",
      "07/31/2016 12:37:55 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:38:21 AM Merging nodes...\n",
      "07/31/2016 12:38:21 AM Done.\n",
      "07/31/2016 12:38:21 AM Done (97.82% identity, 4118561 bases out of 4210369 aligned, 6882 nodes out of 13774 aligned).\n",
      "07/31/2016 12:38:21 AM Writing graph...\n",
      "07/31/2016 12:38:23 AM Done.\n",
      "07/31/2016 12:38:23 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000001F.gfa\n",
      "07/31/2016 12:38:26 AM Reading graph: 000002F.gfa ...\n",
      "07/31/2016 12:38:26 AM Done.\n",
      "07/31/2016 12:38:26 AM Constructing index...\n",
      "07/31/2016 12:38:52 AM Done.\n",
      "07/31/2016 12:38:52 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:39:19 AM Merging nodes...\n",
      "07/31/2016 12:39:19 AM Done.\n",
      "07/31/2016 12:39:19 AM Done (98.51% identity, 4112009 bases out of 4174381 aligned, 5424 nodes out of 10941 aligned).\n",
      "07/31/2016 12:39:19 AM Writing graph...\n",
      "07/31/2016 12:39:21 AM Done.\n",
      "07/31/2016 12:39:21 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000002F.gfa\n",
      "07/31/2016 12:39:24 AM Reading graph: 000003F.gfa ...\n",
      "07/31/2016 12:39:24 AM Done.\n",
      "07/31/2016 12:39:24 AM Constructing index...\n",
      "07/31/2016 12:39:50 AM Done.\n",
      "07/31/2016 12:39:50 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:40:20 AM Merging nodes...\n",
      "07/31/2016 12:40:20 AM Done.\n",
      "07/31/2016 12:40:20 AM Done (96.61% identity, 3533058 bases out of 3657048 aligned, 7155 nodes out of 14852 aligned).\n",
      "07/31/2016 12:40:20 AM Writing graph...\n",
      "07/31/2016 12:40:22 AM Done.\n",
      "07/31/2016 12:40:22 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000003F.gfa\n",
      "07/31/2016 12:40:25 AM Reading graph: 000004F.gfa ...\n",
      "07/31/2016 12:40:25 AM Done.\n",
      "07/31/2016 12:40:25 AM Constructing index...\n",
      "07/31/2016 12:40:51 AM Done.\n",
      "07/31/2016 12:40:51 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:41:11 AM Merging nodes...\n",
      "07/31/2016 12:41:11 AM Done.\n",
      "07/31/2016 12:41:11 AM Done (99.07% identity, 3118950 bases out of 3148353 aligned, 4099 nodes out of 8391 aligned).\n",
      "07/31/2016 12:41:11 AM Writing graph...\n",
      "07/31/2016 12:41:13 AM Done.\n",
      "07/31/2016 12:41:13 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000004F.gfa\n",
      "07/31/2016 12:41:16 AM Reading graph: 000005F.gfa ...\n",
      "07/31/2016 12:41:16 AM Done.\n",
      "07/31/2016 12:41:16 AM Constructing index...\n",
      "07/31/2016 12:41:41 AM Done.\n",
      "07/31/2016 12:41:41 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:42:02 AM Merging nodes...\n",
      "07/31/2016 12:42:02 AM Done.\n",
      "07/31/2016 12:42:02 AM Done (98.49% identity, 2821116 bases out of 2864346 aligned, 6311 nodes out of 11889 aligned).\n",
      "07/31/2016 12:42:02 AM Writing graph...\n",
      "07/31/2016 12:42:04 AM Done.\n",
      "07/31/2016 12:42:04 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000005F.gfa\n",
      "07/31/2016 12:42:07 AM Reading graph: 000006F.gfa ...\n",
      "07/31/2016 12:42:07 AM Done.\n",
      "07/31/2016 12:42:07 AM Constructing index...\n",
      "07/31/2016 12:42:33 AM Done.\n",
      "07/31/2016 12:42:33 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:42:57 AM Merging nodes...\n",
      "07/31/2016 12:42:57 AM Done.\n",
      "07/31/2016 12:42:57 AM Done (98.44% identity, 2771972 bases out of 2815765 aligned, 5141 nodes out of 10482 aligned).\n",
      "07/31/2016 12:42:57 AM Writing graph...\n",
      "07/31/2016 12:42:59 AM Done.\n",
      "07/31/2016 12:42:59 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000006F.gfa\n",
      "07/31/2016 12:43:02 AM Reading graph: 000007F.gfa ...\n",
      "07/31/2016 12:43:02 AM Done.\n",
      "07/31/2016 12:43:02 AM Constructing index...\n",
      "07/31/2016 12:43:28 AM Done.\n",
      "07/31/2016 12:43:28 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:43:54 AM Merging nodes...\n",
      "07/31/2016 12:43:55 AM Done.\n",
      "07/31/2016 12:43:55 AM Done (96.88% identity, 2508959 bases out of 2589667 aligned, 4807 nodes out of 9151 aligned).\n",
      "07/31/2016 12:43:55 AM Writing graph...\n",
      "07/31/2016 12:43:56 AM Done.\n",
      "07/31/2016 12:43:56 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000007F.gfa\n",
      "07/31/2016 12:44:00 AM Reading graph: 000008F.gfa ...\n",
      "07/31/2016 12:44:00 AM Done.\n",
      "07/31/2016 12:44:00 AM Constructing index...\n",
      "07/31/2016 12:44:25 AM Done.\n",
      "07/31/2016 12:44:25 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:44:45 AM Merging nodes...\n",
      "07/31/2016 12:44:46 AM Done.\n",
      "07/31/2016 12:44:46 AM Done (98.83% identity, 2517937 bases out of 2547794 aligned, 4300 nodes out of 8709 aligned).\n",
      "07/31/2016 12:44:46 AM Writing graph...\n",
      "07/31/2016 12:44:48 AM Done.\n",
      "07/31/2016 12:44:48 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000008F.gfa\n",
      "07/31/2016 12:44:51 AM Reading graph: 000009F.gfa ...\n",
      "07/31/2016 12:44:51 AM Done.\n",
      "07/31/2016 12:44:51 AM Constructing index...\n",
      "07/31/2016 12:45:16 AM Done.\n",
      "07/31/2016 12:45:16 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:45:34 AM Merging nodes...\n",
      "07/31/2016 12:45:34 AM Done.\n",
      "07/31/2016 12:45:34 AM Done (98.84% identity, 2477424 bases out of 2506515 aligned, 3174 nodes out of 6490 aligned).\n",
      "07/31/2016 12:45:34 AM Writing graph...\n",
      "07/31/2016 12:45:36 AM Done.\n",
      "07/31/2016 12:45:36 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000009F.gfa\n",
      "07/31/2016 12:45:39 AM Reading graph: 000010F.gfa ...\n",
      "07/31/2016 12:45:39 AM Done.\n",
      "07/31/2016 12:45:39 AM Constructing index...\n",
      "07/31/2016 12:46:04 AM Done.\n",
      "07/31/2016 12:46:04 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:46:31 AM Merging nodes...\n",
      "07/31/2016 12:46:31 AM Done.\n",
      "07/31/2016 12:46:31 AM Done (98.23% identity, 2200069 bases out of 2239622 aligned, 4728 nodes out of 9714 aligned).\n",
      "07/31/2016 12:46:31 AM Writing graph...\n",
      "07/31/2016 12:46:33 AM Done.\n",
      "07/31/2016 12:46:33 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000010F.gfa\n",
      "07/31/2016 12:46:36 AM Reading graph: 000011F.gfa ...\n",
      "07/31/2016 12:46:36 AM Done.\n",
      "07/31/2016 12:46:36 AM Constructing index...\n",
      "07/31/2016 12:47:02 AM Done.\n",
      "07/31/2016 12:47:02 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:47:16 AM Merging nodes...\n",
      "07/31/2016 12:47:16 AM Done.\n",
      "07/31/2016 12:47:16 AM Done (99.24% identity, 2084785 bases out of 2100685 aligned, 2615 nodes out of 5187 aligned).\n",
      "07/31/2016 12:47:16 AM Writing graph...\n",
      "07/31/2016 12:47:18 AM Done.\n",
      "07/31/2016 12:47:18 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000011F.gfa\n",
      "07/31/2016 12:47:21 AM Reading graph: 000012F.gfa ...\n",
      "07/31/2016 12:47:21 AM Done.\n",
      "07/31/2016 12:47:21 AM Constructing index...\n",
      "07/31/2016 12:47:46 AM Done.\n",
      "07/31/2016 12:47:46 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:48:04 AM Merging nodes...\n",
      "07/31/2016 12:48:04 AM Done.\n",
      "07/31/2016 12:48:04 AM Done (99.42% identity, 2037049 bases out of 2048852 aligned, 3049 nodes out of 5846 aligned).\n",
      "07/31/2016 12:48:04 AM Writing graph...\n",
      "07/31/2016 12:48:06 AM Done.\n",
      "07/31/2016 12:48:06 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000012F.gfa\n",
      "07/31/2016 12:48:08 AM Reading graph: 000013F.gfa ...\n",
      "07/31/2016 12:48:09 AM Done.\n",
      "07/31/2016 12:48:09 AM Constructing index...\n",
      "07/31/2016 12:48:34 AM Done.\n",
      "07/31/2016 12:48:34 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:48:50 AM Merging nodes...\n",
      "07/31/2016 12:48:50 AM Done.\n",
      "07/31/2016 12:48:50 AM Done (99.22% identity, 2008364 bases out of 2024130 aligned, 2624 nodes out of 5391 aligned).\n",
      "07/31/2016 12:48:50 AM Writing graph...\n",
      "07/31/2016 12:48:52 AM Done.\n",
      "07/31/2016 12:48:52 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000013F.gfa\n",
      "07/31/2016 12:48:55 AM Reading graph: 000014F.gfa ...\n",
      "07/31/2016 12:48:55 AM Done.\n",
      "07/31/2016 12:48:55 AM Constructing index...\n",
      "07/31/2016 12:49:20 AM Done.\n",
      "07/31/2016 12:49:20 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:49:37 AM Merging nodes...\n",
      "07/31/2016 12:49:37 AM Done.\n",
      "07/31/2016 12:49:37 AM Done (99.23% identity, 2002751 bases out of 2018298 aligned, 2528 nodes out of 5156 aligned).\n",
      "07/31/2016 12:49:37 AM Writing graph...\n",
      "07/31/2016 12:49:39 AM Done.\n",
      "07/31/2016 12:49:39 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000014F.gfa\n",
      "07/31/2016 12:49:42 AM Reading graph: 000015F.gfa ...\n",
      "07/31/2016 12:49:42 AM Done.\n",
      "07/31/2016 12:49:42 AM Constructing index...\n",
      "07/31/2016 12:50:07 AM Done.\n",
      "07/31/2016 12:50:07 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:50:24 AM Merging nodes...\n",
      "07/31/2016 12:50:24 AM Done.\n",
      "07/31/2016 12:50:24 AM Done (98.18% identity, 1916135 bases out of 1951706 aligned, 2320 nodes out of 4731 aligned).\n",
      "07/31/2016 12:50:24 AM Writing graph...\n",
      "07/31/2016 12:50:26 AM Done.\n",
      "07/31/2016 12:50:26 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000015F.gfa\n",
      "07/31/2016 12:50:29 AM Reading graph: 000016F.gfa ...\n",
      "07/31/2016 12:50:29 AM Done.\n",
      "07/31/2016 12:50:29 AM Constructing index...\n",
      "07/31/2016 12:50:54 AM Done.\n",
      "07/31/2016 12:50:54 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:51:13 AM Merging nodes...\n",
      "07/31/2016 12:51:13 AM Done.\n",
      "07/31/2016 12:51:13 AM Done (99.24% identity, 1921427 bases out of 1936231 aligned, 2920 nodes out of 6006 aligned).\n",
      "07/31/2016 12:51:13 AM Writing graph...\n",
      "07/31/2016 12:51:15 AM Done.\n",
      "07/31/2016 12:51:15 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000016F.gfa\n",
      "07/31/2016 12:51:18 AM Reading graph: 000017F.gfa ...\n",
      "07/31/2016 12:51:18 AM Done.\n",
      "07/31/2016 12:51:18 AM Constructing index...\n",
      "07/31/2016 12:51:43 AM Done.\n",
      "07/31/2016 12:51:43 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:51:59 AM Merging nodes...\n",
      "07/31/2016 12:51:59 AM Done.\n",
      "07/31/2016 12:51:59 AM Done (99.31% identity, 1889542 bases out of 1902682 aligned, 1882 nodes out of 3889 aligned).\n",
      "07/31/2016 12:51:59 AM Writing graph...\n",
      "07/31/2016 12:52:01 AM Done.\n",
      "07/31/2016 12:52:01 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000017F.gfa\n",
      "07/31/2016 12:52:04 AM Reading graph: 000018F.gfa ...\n",
      "07/31/2016 12:52:04 AM Done.\n",
      "07/31/2016 12:52:04 AM Constructing index...\n",
      "07/31/2016 12:52:29 AM Done.\n",
      "07/31/2016 12:52:29 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:52:43 AM Merging nodes...\n",
      "07/31/2016 12:52:43 AM Done.\n",
      "07/31/2016 12:52:43 AM Done (99.11% identity, 1877808 bases out of 1894637 aligned, 2471 nodes out of 5017 aligned).\n",
      "07/31/2016 12:52:43 AM Writing graph...\n",
      "07/31/2016 12:52:45 AM Done.\n",
      "07/31/2016 12:52:45 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000018F.gfa\n",
      "07/31/2016 12:52:48 AM Reading graph: 000019F.gfa ...\n",
      "07/31/2016 12:52:48 AM Done.\n",
      "07/31/2016 12:52:48 AM Constructing index...\n",
      "07/31/2016 12:53:14 AM Done.\n",
      "07/31/2016 12:53:14 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:53:29 AM Merging nodes...\n",
      "07/31/2016 12:53:29 AM Done.\n",
      "07/31/2016 12:53:29 AM Done (98.08% identity, 1834342 bases out of 1870332 aligned, 3715 nodes out of 7202 aligned).\n",
      "07/31/2016 12:53:29 AM Writing graph...\n",
      "07/31/2016 12:53:31 AM Done.\n",
      "07/31/2016 12:53:31 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000019F.gfa\n",
      "07/31/2016 12:53:34 AM Reading graph: 000020F.gfa ...\n",
      "07/31/2016 12:53:34 AM Done.\n",
      "07/31/2016 12:53:34 AM Constructing index...\n",
      "07/31/2016 12:53:59 AM Done.\n",
      "07/31/2016 12:53:59 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:54:14 AM Merging nodes...\n",
      "07/31/2016 12:54:14 AM Done.\n",
      "07/31/2016 12:54:14 AM Done (98.93% identity, 1827768 bases out of 1847521 aligned, 2397 nodes out of 4956 aligned).\n",
      "07/31/2016 12:54:14 AM Writing graph...\n",
      "07/31/2016 12:54:16 AM Done.\n",
      "07/31/2016 12:54:16 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000020F.gfa\n",
      "07/31/2016 12:54:19 AM Reading graph: 000021F.gfa ...\n",
      "07/31/2016 12:54:19 AM Done.\n",
      "07/31/2016 12:54:19 AM Constructing index...\n",
      "07/31/2016 12:54:44 AM Done.\n",
      "07/31/2016 12:54:44 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:54:57 AM Merging nodes...\n",
      "07/31/2016 12:54:57 AM Done.\n",
      "07/31/2016 12:54:57 AM Done (98.51% identity, 1833142 bases out of 1860928 aligned, 2430 nodes out of 5115 aligned).\n",
      "07/31/2016 12:54:57 AM Writing graph...\n",
      "07/31/2016 12:54:59 AM Done.\n",
      "07/31/2016 12:54:59 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000021F.gfa\n",
      "07/31/2016 12:55:02 AM Reading graph: 000022F.gfa ...\n",
      "07/31/2016 12:55:02 AM Done.\n",
      "07/31/2016 12:55:02 AM Constructing index...\n",
      "07/31/2016 12:55:28 AM Done.\n",
      "07/31/2016 12:55:28 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:55:44 AM Merging nodes...\n",
      "07/31/2016 12:55:44 AM Done.\n",
      "07/31/2016 12:55:44 AM Done (45.42% identity, 778203 bases out of 1713492 aligned, 1065 nodes out of 2434 aligned).\n",
      "07/31/2016 12:55:44 AM Writing graph...\n",
      "07/31/2016 12:55:46 AM Done.\n",
      "07/31/2016 12:55:46 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000022F.gfa\n",
      "07/31/2016 12:55:49 AM Reading graph: 000023F.gfa ...\n",
      "07/31/2016 12:55:49 AM Done.\n",
      "07/31/2016 12:55:49 AM Constructing index...\n",
      "07/31/2016 12:56:14 AM Done.\n",
      "07/31/2016 12:56:14 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:56:29 AM Merging nodes...\n",
      "07/31/2016 12:56:29 AM Done.\n",
      "07/31/2016 12:56:29 AM Done (98.24% identity, 1681455 bases out of 1711656 aligned, 3196 nodes out of 6597 aligned).\n",
      "07/31/2016 12:56:29 AM Writing graph...\n",
      "07/31/2016 12:56:31 AM Done.\n",
      "07/31/2016 12:56:31 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000023F.gfa\n",
      "07/31/2016 12:56:34 AM Reading graph: 000024F.gfa ...\n",
      "07/31/2016 12:56:34 AM Done.\n",
      "07/31/2016 12:56:34 AM Constructing index...\n",
      "07/31/2016 12:56:59 AM Done.\n",
      "07/31/2016 12:56:59 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:57:15 AM Merging nodes...\n",
      "07/31/2016 12:57:15 AM Done.\n",
      "07/31/2016 12:57:15 AM Done (96.48% identity, 1636467 bases out of 1696168 aligned, 2739 nodes out of 5432 aligned).\n",
      "07/31/2016 12:57:15 AM Writing graph...\n",
      "07/31/2016 12:57:17 AM Done.\n",
      "07/31/2016 12:57:17 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000024F.gfa\n",
      "07/31/2016 12:57:20 AM Reading graph: 000025F.gfa ...\n",
      "07/31/2016 12:57:20 AM Done.\n",
      "07/31/2016 12:57:20 AM Constructing index...\n",
      "07/31/2016 12:57:45 AM Done.\n",
      "07/31/2016 12:57:45 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:58:02 AM Merging nodes...\n",
      "07/31/2016 12:58:02 AM Done.\n",
      "07/31/2016 12:58:02 AM Done (98.47% identity, 1576937 bases out of 1601445 aligned, 3009 nodes out of 6139 aligned).\n",
      "07/31/2016 12:58:02 AM Writing graph...\n",
      "07/31/2016 12:58:04 AM Done.\n",
      "07/31/2016 12:58:04 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000025F.gfa\n",
      "07/31/2016 12:58:06 AM Reading graph: 000026F.gfa ...\n",
      "07/31/2016 12:58:06 AM Done.\n",
      "07/31/2016 12:58:06 AM Constructing index...\n",
      "07/31/2016 12:58:32 AM Done.\n",
      "07/31/2016 12:58:32 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:58:45 AM Merging nodes...\n",
      "07/31/2016 12:58:45 AM Done.\n",
      "07/31/2016 12:58:45 AM Done (96.05% identity, 1459823 bases out of 1519888 aligned, 2397 nodes out of 4954 aligned).\n",
      "07/31/2016 12:58:45 AM Writing graph...\n",
      "07/31/2016 12:58:47 AM Done.\n",
      "07/31/2016 12:58:47 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000026F.gfa\n",
      "07/31/2016 12:58:50 AM Reading graph: 000028F.gfa ...\n",
      "07/31/2016 12:58:50 AM Done.\n",
      "07/31/2016 12:58:50 AM Constructing index...\n",
      "07/31/2016 12:59:16 AM Done.\n",
      "07/31/2016 12:59:16 AM Constructing pairwise-alignment...\n",
      "07/31/2016 12:59:31 AM Merging nodes...\n",
      "07/31/2016 12:59:32 AM Done.\n",
      "07/31/2016 12:59:32 AM Done (98.55% identity, 1401451 bases out of 1422005 aligned, 2404 nodes out of 4812 aligned).\n",
      "07/31/2016 12:59:32 AM Writing graph...\n",
      "07/31/2016 12:59:33 AM Done.\n",
      "07/31/2016 12:59:33 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000028F.gfa\n",
      "07/31/2016 12:59:36 AM Reading graph: 000029F.gfa ...\n",
      "07/31/2016 12:59:36 AM Done.\n",
      "07/31/2016 12:59:36 AM Constructing index...\n",
      "07/31/2016 01:00:01 AM Done.\n",
      "07/31/2016 01:00:01 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:00:18 AM Merging nodes...\n",
      "07/31/2016 01:00:18 AM Done.\n",
      "07/31/2016 01:00:18 AM Done (98.33% identity, 1403815 bases out of 1427622 aligned, 2495 nodes out of 5165 aligned).\n",
      "07/31/2016 01:00:18 AM Writing graph...\n",
      "07/31/2016 01:00:20 AM Done.\n",
      "07/31/2016 01:00:20 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000029F.gfa\n",
      "07/31/2016 01:00:23 AM Reading graph: 000031F.gfa ...\n",
      "07/31/2016 01:00:23 AM Done.\n",
      "07/31/2016 01:00:23 AM Constructing index...\n",
      "07/31/2016 01:00:48 AM Done.\n",
      "07/31/2016 01:00:48 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:01:02 AM Merging nodes...\n",
      "07/31/2016 01:01:03 AM Done.\n",
      "07/31/2016 01:01:03 AM Done (99.39% identity, 1194411 bases out of 1201739 aligned, 2062 nodes out of 3893 aligned).\n",
      "07/31/2016 01:01:03 AM Writing graph...\n",
      "07/31/2016 01:01:04 AM Done.\n",
      "07/31/2016 01:01:04 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000031F.gfa\n",
      "07/31/2016 01:01:07 AM Reading graph: 000032F.gfa ...\n",
      "07/31/2016 01:01:07 AM Done.\n",
      "07/31/2016 01:01:07 AM Constructing index...\n",
      "07/31/2016 01:01:33 AM Done.\n",
      "07/31/2016 01:01:33 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:01:47 AM Merging nodes...\n",
      "07/31/2016 01:01:48 AM Done.\n",
      "07/31/2016 01:01:48 AM Done (98.25% identity, 1172079 bases out of 1193009 aligned, 2464 nodes out of 5065 aligned).\n",
      "07/31/2016 01:01:48 AM Writing graph...\n",
      "07/31/2016 01:01:49 AM Done.\n",
      "07/31/2016 01:01:49 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000032F.gfa\n",
      "07/31/2016 01:01:52 AM Reading graph: 000034F.gfa ...\n",
      "07/31/2016 01:01:53 AM Done.\n",
      "07/31/2016 01:01:53 AM Constructing index...\n",
      "07/31/2016 01:02:18 AM Done.\n",
      "07/31/2016 01:02:18 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:02:29 AM Merging nodes...\n",
      "07/31/2016 01:02:29 AM Done.\n",
      "07/31/2016 01:02:29 AM Done (98.60% identity, 1164675 bases out of 1181163 aligned, 1439 nodes out of 2922 aligned).\n",
      "07/31/2016 01:02:29 AM Writing graph...\n",
      "07/31/2016 01:02:31 AM Done.\n",
      "07/31/2016 01:02:31 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000034F.gfa\n",
      "07/31/2016 01:02:34 AM Reading graph: 000035F.gfa ...\n",
      "07/31/2016 01:02:34 AM Done.\n",
      "07/31/2016 01:02:34 AM Constructing index...\n",
      "07/31/2016 01:02:59 AM Done.\n",
      "07/31/2016 01:02:59 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:03:10 AM Merging nodes...\n",
      "07/31/2016 01:03:11 AM Done.\n",
      "07/31/2016 01:03:11 AM Done (99.60% identity, 1106478 bases out of 1110926 aligned, 1357 nodes out of 2826 aligned).\n",
      "07/31/2016 01:03:11 AM Writing graph...\n",
      "07/31/2016 01:03:12 AM Done.\n",
      "07/31/2016 01:03:12 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000035F.gfa\n",
      "07/31/2016 01:03:15 AM Reading graph: 000036F.gfa ...\n",
      "07/31/2016 01:03:15 AM Done.\n",
      "07/31/2016 01:03:15 AM Constructing index...\n",
      "07/31/2016 01:03:40 AM Done.\n",
      "07/31/2016 01:03:40 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:03:56 AM Merging nodes...\n",
      "07/31/2016 01:03:56 AM Done.\n",
      "07/31/2016 01:03:56 AM Done (96.20% identity, 1033093 bases out of 1073906 aligned, 1773 nodes out of 3616 aligned).\n",
      "07/31/2016 01:03:56 AM Writing graph...\n",
      "07/31/2016 01:03:57 AM Done.\n",
      "07/31/2016 01:03:57 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000036F.gfa\n",
      "07/31/2016 01:04:00 AM Reading graph: 000037F.gfa ...\n",
      "07/31/2016 01:04:00 AM Done.\n",
      "07/31/2016 01:04:00 AM Constructing index...\n",
      "07/31/2016 01:04:25 AM Done.\n",
      "07/31/2016 01:04:25 AM Constructing pairwise-alignment...\n",
      "07/31/2016 01:04:40 AM Merging nodes...\n",
      "07/31/2016 01:04:40 AM Done.\n",
      "07/31/2016 01:04:40 AM Done (97.54% identity, 1009468 bases out of 1034970 aligned, 1579 nodes out of 3013 aligned).\n",
      "07/31/2016 01:04:40 AM Writing graph...\n",
      "07/31/2016 01:04:42 AM Done.\n",
      "07/31/2016 01:04:42 AM Alignment graph written to: Caenorhabditis_elegans.WBcel235.31.dna.genome_000037F.gfa\n"
     ]
    }
   ],
   "source": [
    "for gfa in contigs:\n",
    "    !reveal align -m5 Caenorhabditis_elegans.WBcel235.31.dna.genome.fa {gfa}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Most graphs aligned with pretty high identity to the reference assembly. Lets detect bubbles from the resulting graphs that now describe three paths. Two for the two chromosomes and one for the haploid WBcel235 assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "for gfa in contigs:\n",
    "    !reveal bubbles Caenorhabditis_elegans.WBcel235.31.dna.genome_{gfa} -r Caenorhabditis_elegans.WBcel235.31.dna.genome.fa > {gfa.replace(\".gfa\",\".refaligned\")}.bubbles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From these bubbles, we should now be able to detect both, homozygous variants (where both chromosomes deviate from the WBcel235 assembly) and heterozygous variants (where just one of two chromosomes deviates from the reference assembly)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detected 260 variations with more than 2 alleles.\n",
      "Detected 82 complex bubbles structures.\n",
      "Detected 109 heterozygous SNPs.\n",
      "Detected 5635 heterozygous indels.\n",
      "Detected 15 heterozygous structural variants wrt reference.\n",
      "Detected 2220 homozygous SNPs.\n",
      "Detected 92922 homozygous indels.\n",
      "Detected 1259 homozygous structural variants wrt reference.\n"
     ]
    }
   ],
   "source": [
    "hetsnps=[]\n",
    "hetindels=[]\n",
    "hetstruct=[]\n",
    "homsnps=[]\n",
    "homindels=[]\n",
    "homstruct=[]\n",
    "multiallele=0\n",
    "complexb=0\n",
    "for gfa in contigs:\n",
    "    bubblefile=gfa.replace(\".gfa\",\".refaligned.bubbles\")\n",
    "    with open(bubblefile) as bubbles:\n",
    "        header=bubbles.readline()\n",
    "        headers=header.split()\n",
    "        s1,s2,s3=(headers[6],headers[7],headers[8]) #determine which sample corresponds to which genotype\n",
    "        if s1==\"Caenorhabditis_elegans.WBcel235.31.dna.genome.fa\":\n",
    "            refidx=6\n",
    "            altidx1=7\n",
    "            altidx2=8\n",
    "        elif s2==\"Caenorhabditis_elegans.WBcel235.31.dna.genome.fa\":\n",
    "            refidx=7\n",
    "            altidx1=6\n",
    "            altidx2=8\n",
    "        else:\n",
    "            refidx=8\n",
    "            altidx1=6\n",
    "            altidx2=7\n",
    "        \n",
    "        for line in bubbles:\n",
    "            bubble=line.split()\n",
    "            alleles=bubble[5].split(',')\n",
    "            \n",
    "            varlengths=[len(a) for a in alleles]\n",
    "            size=max(varlengths)\n",
    "            snp=set([1])==set(varlengths) and '-' not in alleles and 'N' not in alleles\n",
    "            indel='-' in alleles\n",
    "            intersectsall=bubble[refidx]!='-' and bubble[altidx1]!='-' and bubble[altidx2]!='-'\n",
    "            \n",
    "            if len(alleles)>2:\n",
    "                multiallele+=1\n",
    "                continue #more than 2 alleles, skip\n",
    "            if 'N' in alleles:\n",
    "                complexb+=1\n",
    "                continue #complex bubble\n",
    "            \n",
    "            if snp and intersectsall:\n",
    "                if bubble[altidx1]!=bubble[altidx2]:\n",
    "                    hetsnps.append((bubblefile,line))\n",
    "                else:\n",
    "                    homsnps.append((bubblefile,line))\n",
    "            if indel and intersectsall:\n",
    "                if bubble[altidx1]!=bubble[altidx2]:\n",
    "                    hetindels.append((bubblefile,line))\n",
    "                else:\n",
    "                    homindels.append((bubblefile,line))\n",
    "                    \n",
    "            if size>50 and intersectsall:\n",
    "                if bubble[altidx1]!=bubble[altidx2]:\n",
    "                    hetstruct.append((bubblefile,line))\n",
    "                else:\n",
    "                    homstruct.append((bubblefile,line))\n",
    "\n",
    "print \"Detected %d variations with more than 2 alleles.\"%multiallele\n",
    "print \"Detected %d complex bubbles structures.\"%complexb\n",
    "\n",
    "print \"Detected %d heterozygous SNPs.\"%len(hetsnps)\n",
    "print \"Detected %d heterozygous indels.\"%len(hetindels)\n",
    "print \"Detected %d heterozygous structural variants wrt reference.\"%len(hetstruct)\n",
    "\n",
    "print \"Detected %d homozygous SNPs.\"%len(homsnps)\n",
    "print \"Detected %d homozygous indels.\"%len(homindels)\n",
    "print \"Detected %d homozygous structural variants wrt reference.\"%len(homstruct)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It looks like quite some complex bubble structures occured. Also, we lost some heterozygous variations that we observed in simple bubble structures from the graph based representation. We do also detect homozygous variations, but once again these are dominated by short indels. Although it seems like the approach worked, the amount of short indels is probably too high for our method to come up with proper variant calls."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
